### PR TITLE
util/winutil: add package for logging into Windows via Service-for-Us…

### LINKED ISSUE #12383: Windows: add support for service-for-user (S4U)
S4U is the mechanism by which we can obtain a Windows access token for use by SSH, Taildrop, et al.

### PR DESCRIPTION
**Pull request by dblohm7**
…er (S4U)

This PR ties together pseudoconsoles, user profiles, s4u logons, and process creation into what is (hopefully) a simple API for various Tailscale services to obtain Windows access tokens without requiring knowledge of any Windows passwords. It works both for domain-joined machines (Kerberos) and non-domain-joined machines. The former case is fairly straightforward as it is fully documented. OTOH, the latter case is not documented, though it is fully defined in the C headers in the Windows SDK. The documentation blanks were filled in by reading the source code of Microsoft's Win32 port of OpenSSH.

We need to do a bit of acrobatics to make conpty work correctly while creating a child process with an s4u token; see the doc comments above startProcessInternal for details.

Fixes #12383

### CODE REVIEW

File: util/winutil/s4u/s4u_windows.go
Comment (nickkhyl) [inline util/winutil/s4u/s4u_windows.go:599]: Can you please document that `(*Process).Close` waits for the process to exit and explain the reason behind this (we cannot/should not unload the user's profile until then)?

Also, I believe the implication is that the caller should invoke `(*Process).Terminate` before calling `(*Process).Close`. However, this is not obvious from the method name or doc

**Review by nickkhyl — CHANGES_REQUESTED** [review]
That's very cool, and I think we should be good to go with the comments addressed!
That being said, I'd like to sleep on it and take a second look before it gets merged.

File: util/winutil/s4u/lsa_windows.go
Comment (dblohm7) [reply util/winutil/s4u/lsa_windows.go:259]: I filed https://github.com/tailscale/tailscale/issues/12561 for this; it feels to me like we need to have a broader conversation about how we deal with this situation.

File: util/winutil/s4u/s4u_windows.go
Comment (dblohm7) [reply util/winutil/s4u/s4u_windows.go:38]: I originally named those values `Identification` and `Impersonation`, however as I fleshed out the API I realized that I didn't really want to expose a bunch of Windows NT concepts to the exported API. I'm imagining the scenario where somebody is working on a new feature (say, TailDrive) and wants to use S4U for something. I want these values to be meaningful and clear, but also understandable to somebody who isn't necessarily familiar with the concepts of identification vs impersonation.

I'm welcome to suggestions that could improve that clarity -- confusion is definitely something I want to avoid!

Comment (nickkhyl) [reply util/winutil/s4u/s4u_windows.go:38]: I see your point, and I DO like the higher-level `s4u.Login` / `(*s4u.Session).Do` / `(*s4u.Session).StartProcess` set of API. I also expect it to be sufficient for most, if not all, scenarios. And we might want to provide higher-level APIs for some of the remaining scenarios.

For example, someone might need to access user's environment variables, or expand them in a given string. But in order for, say, `windows.Token.Environ` to work correctly and return user-specific variables, user's profile must be loaded. Additionally, and that's probably undocumented, some of the variables will only be set in the newly created environment block if the token is an impersonation (rather than identification) token. Therefore, I think it might be beneficial to allow s4u package users to access `s4u.Session`'s token and/or provide higher-level APIs that provide access to user's environment variables.

I do not oppose having the lower-level `s4u.CreateToken` API as exported. However I believe that its users must be familiar with the Windows NT concepts, and have a clear understanding of what it returns, and what it can and cannot be used for. From this perspective, I think it's better to make the names clear and searchable. And if we feel like we need to make it easier to use for non-Windows folks, maybe we could guide them in the documentation? Saying something like "An identification token can be used to query information about the user, while an impersonation token allows tailscaled to act on behalf of the user. CreateToken does not load the user's profile and cannot be used for operations like accessing the user's registry key or expanding user-specific environment variables. Consider using s4u.Login instead."

File: util/winutil/s4u/s4u_windows.go
Comment (dblohm7) [reply util/winutil/s4u/s4u_windows.go:100]: Yep, I'll follow-up after this one merges.

File: util/winutil/s4u/s4u_windows.go
Comment (dblohm7) [reply util/winutil/s4u/s4u_windows.go:38]: That makes a lot of sense. I actually expose `CreateToken` because Tailscale SSH wants to query group memberships prior to logging in... Maybe the best way to deal with this is to make `CreateToken` internal and just expose the group-based API directly.

File: util/winutil/s4u/s4u_windows.go
Comment (dblohm7) [reply util/winutil/s4u/s4u_windows.go:367]: As an alternative, what if we set the process mitigation policy forcing system32 to the front of the search path?

Comment (nickkhyl) [reply util/winutil/s4u/s4u_windows.go:367]: One related issue that I believe the process mitigation policy won't address is a malicious executable with a well-known name placed by a lower-integrity process into the user's profile directory. For example, if there is a malicious `ipconfig.bat`/`ipconfig.exe` in `%userprofile%`, and the user types `ipconfig` via SSH, that's what will run instead of the system utility.

One could argue that they could change into their home directory themselves before running `ipconfig` and would be right. But I'm not sure if we should make it easier on malware.

We could also mitigate this attack vector by using the limited admin's token. But this would also make the whole feature less enticing.

File: util/winutil/s4u/lsa_windows.go
Comment (dblohm7) [reply util/winutil/s4u/lsa_windows.go:259]: I just used literal `0` given `Token`'s underlying type.

File: util/winutil/s4u/s4u_windows.go
Comment (dblohm7) [reply util/winutil/s4u/s4u_windows.go:367]: Ah, right. Will fix.

**Review by nickkhyl — APPROVED** [review]
LGTM with case-insensitive environment variables sorting and deduplication.
I'd also prefer a safer, harder-to-misuse alternative to `s4u.ListGroupIDs` and would like to learn more about how we're going to use it, but I won't block on this.

File: util/winutil/s4u/s4u_windows.go
Comment (dblohm7) [reply util/winutil/s4u/s4u_windows.go:54]: The Windows implementation of `(*userMeta).GroupIds` (see https://github.com/tailscale/tailscale/blob/main/ssh/tailssh/user.go).

File: util/winutil/s4u/s4u_windows.go
Comment (dblohm7) [reply util/winutil/s4u/s4u_windows.go:473]: > Can you please make it case-insensitive?
> Also, if we choose to do it here rather than in `winutil.NewEnvBlock`, I think we should at least update the `winutil.NewEnvBlock` doc to say that variables in `env` must be sorted by names (case-insensitive).

Sure, that sounds reasonable.

> Additionally, I think we should handle cases when `env` already contains variables with the same (case-insensitive) names as in `extraEnv`.

That's why I used a stable sort: the newer entry will be ordered after any extant entry with the same name.

File: util/winutil/s4u/s4u_windows.go
Comment (dblohm7) [reply util/winutil/s4u/s4u_windows.go:54]: @nickkhyl How about I rename it something like `ListGroupIDsForSSHOnly` and add additional commentary to the doc string?

File: util/winutil/s4u/s4u_windows.go
Comment (nickkhyl) [reply util/winutil/s4u/s4u_windows.go]: >That's why I used a stable sort: the newer entry will be ordered after any extant entry with the same name.
Please correct me if I'm wrong, but this sounds like you're expecting variables that come later in the environment block to replace/win over variables defined earlier. The thing is, that's not how it works; a valid environment block should not contain any duplicate variables.

File: util/winutil/s4u/s4u_windows.go
Comment (nickkhyl) [reply util/winutil/s4u/s4u_windows.go:57]: > How about I rename it something like ListGroupIDsForSSHOnly and add additional commentary to the doc string?

Sounds good to me.

> The Windows implementation of (*userMeta).GroupIds

Thank you!

File: util/winutil/s4u/s4u_windows.go
Comment (nickkhyl) [inline util/winutil/s4u/s4u_windows.go:915]: Nice, thank you!

**Review by nickkhyl — APPROVED** [review]

File: util/winutil/s4u/s4u_windows.go
Comment (dblohm7) [reply util/winutil/s4u/s4u_windows.go]: Filed https://github.com/tailscale/tailscale/issues/12627

**Pull request merged**
